ang/civiawards-workflow/services/applicant-management-workflow.service.js

```diff
@@ -0,0 +1,164 @@
+    ContactsCache, processMyAwardsFilter, Select2Utils, isTruthy) {
+    var awardSubtypes = AwardSubtype.getAll();
+
+    this.createDuplicate = createDuplicate;
+    this.getWorkflowsList = getWorkflowsList;
+
+    /**
+     * @param {object} selectedFilters selected filter values
+     * @returns {Promise} Promise which resolved to a list of case type ids
+     */
+    function fetchFilteredWorkflowIds (selectedFilters) {
+      return processMyAwardsFilter(selectedFilters.awardFilter)
+        .then(function (caseTypeIDs) {
+          if (caseTypeIDs.length === 0) {
+            return $q.resolve([]);
+          }
+
+          var filters = prepareFilterObject(selectedFilters, caseTypeIDs);
+
+          return civicaseCrmApi('AwardDetail', 'get', filters)
+            .then(function (awardsData) {
+              return _.map(awardsData.values, function (award) {
+                return award.case_type_id;
+              });
+            });
+        });
+    }
+
+    /**
+     * Returns workflows list for applicant management
+     *
+     * @param {object} caseTypeCategoryName case type category name
+     * @param {object} selectedFilters selected filter values
+     * @returns {Array} api call parameters
+     */
+    function getWorkflowsList (caseTypeCategoryName, selectedFilters) {
+      return fetchFilteredWorkflowIds(selectedFilters)
+        .then(function (workflowIds) {
+          if (workflowIds.length === 0) {
+            return $q.resolve([]);
+          }
+
+          var params = {
+            id: { IN: workflowIds },
+            sequential: 1,
+            is_active: selectedFilters.is_active,
+            case_type_category: caseTypeCategoryName,
+            options: { limit: 0 },
+            'api.AwardDetail.get': {
+              case_type_id: '$value.id'
+            }
+          };
+
+          return civicaseCrmApi('CaseType', 'get', params).then(function (data) {
+            return getFormattedAwardDetailsData(data.values);
+          });
+        });
+    }
+
+    /**
+     * @param {object} workflow workflow object
+     * @returns {Promise} promise
+     */
+    function createDuplicate (workflow) {
+      return civicaseCrmApi('CaseType', 'create', _.extend({}, workflow, { id: null }))
+        .then(function (data) {
+          var awardsDetailsData = _.clone(workflow.awardDetails);
+
+          awardsDetailsData.id = null;
+          awardsDetailsData.case_type_id = data.values[0].id;
+
+          return civicaseCrmApi([
+            ['AwardDetail', 'create', awardsDetailsData]
+          ]);
+        });
+    }
+
+    /**
+     * @param {object[]} workflows list of workflows
+     * @returns {object[]} formatted list of workflows
+     */
+    function getFormattedAwardDetailsData (workflows) {
+      var managerContacts = getAwardManagersForWorkflows(workflows);
+
+      return ContactsCache.add(managerContacts)
+        .then(function () {
+          var workflowsCopy = _.clone(workflows);
```

Comment:
`clone` only clones the array, not the objects inside. You need to use `cloneDeep` if you want this to work without mutating the original workflows.

I think a better approach here is to use map:

```js
return _.map(workflows, function (workflow) {
  var workflowCopy = _.clone(workflow);
  // ...
  return workflowCopy;
});
```
I think we should start using `map` more. Map means that you are transforming the values in a array into a new value. Formatting workflows is doing just that and by using `map` it makes things more intuitive. `each` is supposed to use to walk through an array and we should use it when we are not expecting to return a value. For example, if we were to call a list of actions `each` would be better.

Reply:
you can also use `pick` to get the attributes you expect from an object and ignore the rest:

```js
_.pick({ a: 1, b: 2, c: 3}, ['a', 'c']); // { a: 1, c: 3 }
```

Instead of cloning we can do the following:

```js
return _.map(workflows, function (workflow) {
  return _.assign(
    {},
    _.pick(workflow, ['attribute1', 'attribute2']),
    {
      awardDetails: ...,
      awardDetailsFormatted: { ... }
    }
});
```

And if the list of attributes is too long and you rather just omit a certain property you can use `omit`:

```js
_.omit({ a: 1, b: 2, c: 3}, ['b']); // { a: 1, c: 3 }
```